src/Mutts.jl

```diff
@@ -23,41 +27,61 @@ TODO
   immutable.
 =#
 
-export Mutt, @mutt, branch!, ismutable, markimmutable!, getmutableversion!
+export @mutt, ismuttstype, branch!, ismutable, markimmutable!, getmutableversion!
 
 """
-    abstract type Mutt end
-Types created via `@mutt`. This means they implement the _mutable-until-shared_ discipline.
-"""
-abstract type Mutt end
-
-ismutable(obj :: Mutt) = obj.__mutt_mutable
-Base.isimmutable(obj :: Mutt) = !ismutable(obj)
+    ismuttstype(t::Type) -> Bool
 
-function getmutableversion!(obj :: Mutt)
+Returns true if type `t` was created via the `@mutt` macro, meaning it is a Mutable Til
+Shared type, and implements the _mutable-until-shared_ discipline. These types start out
+mutable, and can be saved in a frozen version via `markimmutable!` and `branch!`.
+"""
+ismuttstype(t::Type) = mutts_trait(t) == MuttsType()
+
+# --- Trait dispatch for Mutt types -----------------------
+# Since Mutts types don't inherit from a common abstract type (to allow them to inherit
+# from their own base types, e.g. AbstractDict), we use trait-inheritance to write functions
+# that want different methods for Mutts and non-Mutts types.
+# (For more on Traits in Julia ("Holy Traits"), see this blog post):
+# https://invenia.github.io/blog/2019/11/06/julialang-features-part-2/
+struct MuttsType end
+struct NonMuttsType end
+# Types are consider non-Mutts by default, and the @mutt macro will create an overload for
+# `mutts_trait(T)` marking them as MuttsType.
+mutts_trait(::Type) = NonMuttsType()
+mutts_trait(v) = mutts_trait(typeof(v))
```

Comment:
I'm not sure about the performance aspects of how to invoke the trait.

In Lyndon's blog post (https://invenia.github.io/blog/2019/11/06/julialang-features-part-2) he uses:
```julia
ismutable(obj::T) where T = ismutable(mutts_trait(T), obj)
```
exactly as I've done for `ismutable`, below.

But other options that seem reasonable to me are to use this value overload above, and write the trait dispatching functions like this:
```julia
ismutable(obj) = ismutable(mutts_trait(obj), obj)
```
And/or to use the `typeof` explictly like this:
```julia
ismutable(obj) = ismutable(mutts_trait(typeof(obj)), obj)
```
Probably those last two are equivalent.

I worry about having these `where T`s everywhere resulting in unneeded excess specialization, which slows down compilation. On the other hand, one of the cool things about these Holy Traits is that they compile away, and maybe you _need_ that specialization in order to get them to do that?

I'll ask Lyndon and see what he says! :)

Reply:
Okay, nvm, after thinking more about this, there's nothing wrong with the first version, since it will be specialized anyway (the argument is a value, not a Type).

That said, we might actually want to consider marking `mutts_trait(::Type)` and `mutts_trait(v)` as `@nospecialize`, since they'll be identical for all types. But before I do that, i'd want to make sure that doesn't disable the optimizations. (I think it wouldn't, but not 100% sure.)